JPEG XS essence support

nishabfhg opened the pull request:
Hi, 

Here are my changes to include JPEG XS essence support into bmx. It works pretty much like JPEG 2000. Only PARSE_FRAME_START_SIZE has to be larger in RawEssenceReader. Let me know if I need to modify something.  

Thank you.
Best regards, 
Nisha

philipnbbc commented:
Thanks! I don't have time available today to review but will try make a start tomorrow or Wednesday. Is there a sample available somewhere online I can use to test?

nishabfhg commented:
Great! I don't know if there are any jxs samples available online. I have added a short RGB test sequence here: https://github.com/nishabfhg/bmx-jxs/tree/main/Blender_FullHD_Testsequence. In raw2bmx, you will need the jxs_rgba command line option in this case.

philipnbbc commented on deps/libMXFpp/libMXF++/metadata/base/JPEGXSSubDescriptorBase.cpp:
```diff
@@ -0,0 +1,155 @@
+
+std::vector<uint8_t> JPEGXSSubDescriptorBase::getJPEGXSComponentTable() const
+{
+	return getUInt8ArrayItem(&MXF_ITEM_K(JPEGXSSubDescriptor, JPEGXSComponentTable));
```
This should use `getRawBytesItem`

philipnbbc commented on deps/libMXFpp/libMXF++/metadata/base/JPEGXSSubDescriptorBase.cpp:
```diff
@@ -0,0 +1,155 @@
+
+void JPEGXSSubDescriptorBase::setJPEGXSComponentTable(std::vector<uint8_t> value)
+{
+	setUInt8JxsArray(&MXF_ITEM_K(JPEGXSSubDescriptor, JPEGXSComponentTable), value);
```
This should use `setRawBytesItem`.

philipnbbc commented on include/bmx/essence_parser/FilePatternEssenceSource.h:
```diff
@@ -50,6 +50,7 @@ class FilePatternEssenceSource : public EssenceSource
 
 public:
     virtual uint32_t Read(unsigned char *data, uint32_t size);
+	int64_t ReadFileSize(uint32_t index);
```
This is unused. Can it be removed?

philipnbbc commented on src/common/EssenceType.cpp:
```diff
@@ -117,6 +117,8 @@ static const EssenceTypeInfo ESSENCE_TYPE_INFO[] =
     {RDD36_4444_XQ,             PICTURE_ESSENCE,        "RDD36 4:4:4:4 XQ",                     "RDD36_4444_XQ"},
     {JPEG2000_CDCI,             PICTURE_ESSENCE,        "JPEG 2000 CDCI",                       "JPEG2000_CDCI"},
     {JPEG2000_RGBA,             PICTURE_ESSENCE,        "JPEG 2000 RGBA",                       "JPEG2000_RGBA"},
+	{JPEGXS_CDCI,               PICTURE_ESSENCE,        "JPEG XS CDCI",                         "JPEGXS_CDCI"},
+	{JPEGXS_RGBA,               PICTURE_ESSENCE,        "JPEG XS RGBA",                         "JPEGXS_RGBA"},
```
Could you also add entries to the XML enum around [here](https://github.com/bbc/bmx/blob/main/meta/mxf2raw/mxf2raw_info.xsd#L981)?

philipnbbc commented on src/common/EssenceType.cpp:
```diff
@@ -167,7 +169,7 @@ const char* bmx::essence_type_to_enum_string(EssenceType essence_type)
 {
     BMX_ASSERT((size_t)essence_type < BMX_ARRAY_SIZE(ESSENCE_TYPE_INFO));
-    BMX_ASSERT(ESSENCE_TYPE_INFO[essence_type].essence_type == essence_type);
+    //BMX_ASSERT(ESSENCE_TYPE_INFO[essence_type].essence_type == essence_type);
```
Assume this can be uncommented now?

philipnbbc commented on deps/libMXF/mxf/mxf_labels_and_keys.h:
```diff
@@ -327,6 +327,24 @@ static const mxfUL MXF_CMDEF_L(JPEG2000_HTJ2K_GENERIC) = MXF_JPEG2000_CMDEV_L(MX
+static const mxfUL MXF_CMDEF_L(JPEGXSHigh444_12Profile)			= MXF_JPEGXS_CMDEV_L(0x08);
+static const mxfUL MXF_CMDEF_L(JPEGXSHigh4444_12Profile)		= MXF_JPEGXS_CMDEV_L(0x09);
+
+//void mxf_get_jpegxs_coding_label(mxfUL *label);
```
Can be removed?

philipnbbc reviewed (commented):
I've only managed to review some of it thus far but it is looking good. I'm certain it hasn't been trivial to figure out where things need to go!

There are a couple comments inline. Beyond that you need to convert the tabs to (4) spaces :-) There are also some build warnings which I'll copy into a comment as I don't think you can see the build results in the GitHub Actions output.

I'll try and complete the review this week.

philipnbbc commented:
Here is the list of warnings from the Linux build:
```
/home/runner/work/bmx/bmx/src/essence_parser/FilePatternEssenceSource.cpp: In member function ‘int64_t bmx::FilePatternEssenceSource::ReadFileSize(uint32_t)’:
/home/runner/work/bmx/bmx/src/essence_parser/FilePatternEssenceSource.cpp:208:14: warning: unused variable ‘res’ [-Wunused-variable]
  208 |         bool res = NextFile();
      |              ^~~
```

```
In file included from /home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:33:
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp: In member function ‘virtual void bmx::JXSEssenceParser::ParseFrameInfo(const unsigned char*, uint32_t)’:
/home/runner/work/bmx/bmx/include/bmx/essence_parser/JXSEssenceParser.h:49:34: warning: comparison of unsigned expression in ‘< 0’ is always false [-Wtype-limits]
   49 |         #define SUCCESS(v) (((v) < 0) ? 0 : 1)
      |                              ~~~~^~~
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:144:29: note: in expansion of macro ‘SUCCESS’
  144 |         while (p < end_p && SUCCESS(result))
      |                             ^~~~~~~
/home/runner/work/bmx/bmx/include/bmx/essence_parser/JXSEssenceParser.h:50:34: warning: comparison of unsigned expression in ‘< 0’ is always false [-Wtype-limits]
   50 |         #define FAILURE(v) (((v) < 0) ? 1 : 0)
      |                              ~~~~^~~
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:148:21: note: in expansion of macro ‘FAILURE’
  148 |                 if (FAILURE(result))
      |                     ^~~~~~~
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:211:66: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  211 |                                 if (PIH_.Hsl() < 1 || PIH_.Hsl() > 0xffff) // This includes the EOF check
      |                                                       ~~~~~~~~~~~^~~~~~~~
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:154:24: warning: enumeration value ‘MRK_NIL’ not handled in switch [-Wswitch]
  154 |                 switch (NextMarker.m_Type)
      |                        ^
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:154:24: warning: enumeration value ‘MRK_WGT’ not handled in switch [-Wswitch]
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:154:24: warning: enumeration value ‘MRK_COM’ not handled in switch [-Wswitch]
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:154:24: warning: enumeration value ‘MRK_NLT’ not handled in switch [-Wswitch]
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:154:24: warning: enumeration value ‘MRK_CWD’ not handled in switch [-Wswitch]
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:154:24: warning: enumeration value ‘MRK_CTS’ not handled in switch [-Wswitch]
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:154:24: warning: enumeration value ‘MRK_CRG’ not handled in switch [-Wswitch]
/home/runner/work/bmx/bmx/src/essence_parser/JXSEssenceParser.cpp:154:24: warning: enumeration value ‘MRK_CAP’ not handled in switch [-Wswitch]
```

```
/home/runner/work/bmx/bmx/apps/bmxtranswrap/bmxtranswrap.cpp: In function ‘int main(int, const char**)’:
/home/runner/work/bmx/bmx/apps/bmxtranswrap/bmxtranswrap.cpp:3967:20: warning: enumeration value ‘JPEGXS_CDCI’ not handled in switch [-Wswitch]
 3967 |             switch (output_essence_type)
      |                    ^
/home/runner/work/bmx/bmx/apps/bmxtranswrap/bmxtranswrap.cpp:3967:20: warning: enumeration value ‘JPEGXS_RGBA’ not handled in switch [-Wswitch]
```

philipnbbc commented:
As promised, here is the process for creating regression tests for JXS.

More info on testing can be found [here](https://github.com/bbc/bmx/blob/main/docs/build.md#testing). This process assumes you only have an RGBA example. If you also have a CDCI example file then ignore the steps below that delete that particular part of the tests; you will have 4 test data files instead of 2.

* Copy `tests/jpeg2000` to `tests/jpegxs`
* Remove the test inputs and data: .j2c .bin and .md5 files
* Copy in file 00000 and 00003 from `Blender_FullHD_Testsequence/` (i.e. leave a gap which will be filled in using the `--fill-pattern-gaps` option in the tests)
* Modify `tests/CMakeLists.txt` 
	* Add a  `add_subdirectory(jpegxs)` after the jpeg2000 inclusion
* Modify `.gitattributes`
	* to have only 2 lines for `_1` and `_2`. These test data files will be generated later
* Modify `tests/jpegxs/CMakeLists.txt`
	* Remove `cdci` from the tests list
	* Replace `jpeg2000` with `jpegxs`
* Modify `tests/jpegxs/test_rgba.cmake`
	* Change the `_3` and `_4` in names to `_1` and  `_2`
	* Change the `--j2c_rgba` arg to `--jxs_rgba "${TEST_SOURCE_DIR}/Blender_Sequence_FullHD_1920x1080_8b_444_%d.jxs"`
	* Remove `--aspect-ratio` and `--frame-layout` (which will be parsed from the bitstream) and other args like `--transfer-ch` as required that don't match the content
* Rebuild using `cmake --build .`
* Generate sample files for the test using `cmake --build . --target bmx_jpegxs_rgba_data`
	* the files will be output to the `test_samples/jpegxs/` sub-directory in the build directory
	* Check that the sample files look ok. The .bin files  are the XML output from mxf2raw and the .mxf files are generated by raw2bmx and bmxtranswrap.
* Generate the test data (.bin and .md5) files using `cmake --build . --target bmx_jpegxs_rgba_data`
* Run the tests using `ctest` to check the new JXS tests are run and pass
* Remove the `Blender_FullHD_Testsequence` directory

philipnbbc commented on src/mxf_helper/JPEGXSMXFDescriptorHelper.cpp:
```diff
@@ -0,0 +1,363 @@
+	mJPEGXSSubDescriptor->setJPEGXSNc(essence_parser->GetJPEGXSNc());
+	mJPEGXSSubDescriptor->setJPEGXSHsl(essence_parser->GetJPEGXSHsl());
+
+	int length = essence_parser->GetJPEGXSComponentTable().GetSize();
+	std::vector<uint8_t> tmp(length);
+	ByteArray tmpArr = essence_parser->GetJPEGXSComponentTable();
+	unsigned char* data = (tmpArr).GetBytes();
+	for (int i = 0; i < length; i++)
+	{
+		tmp[i] = (*data);
+		data++;
+	}
+	
+	// for loop each element in the data pointer to the array
+	mJPEGXSSubDescriptor->setJPEGXSComponentTable(tmp);
```
Could this just be something like this?
```
start = essence_parser->GetJPEGXSComponentTable().GetBytes();
end = start + essence_parser->GetJPEGXSComponentTable().GetSize());
SetJPEGXSComponentTable(vector<uint8_t>(start, end));
```
This will set the property in the sub-descriptor and set `mComponentTable` which `UpdateFileDescriptor` will use to set the pixel layout. See also the comment below about the pixel layout being overriden.

philipnbbc reviewed (commented):
I've completed the review and further comments / suggestions can be found inline

nishabfhg commented:
Hi Philip, thanks for the review! I will try to make all the modifications next week and get back to you.

nishabfhg commented on include/bmx/essence_parser/JXSEssenceParser.h:
```diff
@@ -0,0 +1,337 @@
+			{
+				assert(M.m_Type == MRK_CAP);
+				m_MarkerData = M.m_Data;
+				m_DataSize = M.m_DataSize - 2;
```
You are right

nishabfhg commented on src/essence_parser/RawEssenceReader.cpp:
```diff
@@ -53,9 +53,10 @@ using namespace std;
+//#define READ_BLOCK_SIZE         8192
+//#define PARSE_FRAME_START_SIZE  8192
+#define READ_BLOCK_SIZE         3000000
+#define PARSE_FRAME_START_SIZE  3000000   // 156556800
```
Yes, this was an optimization for JPEG XS since the frame size is in general much larger. I have included another setter for this and left 8192 as default.

nishabfhg commented on src/essence_parser/JXSEssenceParser.cpp:
```diff
@@ -0,0 +1,302 @@
+				}
+					
+				// size of the bitstream: ignore, just store away.
+				size = PIH_.LcodSize();
```
This wouldn't work when ParseFrameSize() is called with the default PARSE_FRAME_START_SIZE/ mFrameStartSize.

nishabfhg commented:
Hi Philip, 
I have modified the code according to your review points. I also added a regression test for jpegxs cdci. Let me know if there's anything else. (I will only be able to get back to this after the New Year holidays)

philipnbbc commented:
I've had a quick look through and looks like you covered everything, thanks for the updates! I'll probably take a proper look sometime next week as I'm away from work.

philipnbbc reviewed (commented):
Thanks for your update!. There are a couple remaining issues to look at (most of the comments are whitespace related).

There are also a couple of merge conflicts that have arisen because of recent updates to main.

The test data is likely to change and you can run `cmake --build . --target bmx_jpegxs_rgba_data` and `cmake --build . --target bmx_jpegxs_cdci_data` to update them.

nishabfhg commented:
Hi Philip. I made the fixes. I hope I didn't miss anything. I synced my version with your latest commits, so the merge conflicts should not appear anymore.

philipnbbc commented:
All done now except for some unexpected changes to some non-jxs test checksums - see below the diff showing what needs to be changed. Do you know what caused this? Just asking as it might be an issue that needs fixing.

If you run `cmake --build . -t bmx_test_data` and `cmake --build . -t libMXF_test_data` it should regenerate the checksums as bbelow. `cmake --build . -t test` should then pass.

Diff showing what needs to be changed back:
```
diff --git a/deps/libMXF/test/test_indextable.md5 b/deps/libMXF/test/test_indextable.md5
index e781305e..6dc5d47a 100644
--- a/deps/libMXF/test/test_indextable.md5
+++ b/deps/libMXF/test/test_indextable.md5
@@ -1 +1 @@
-d41d8cd98f00b204e9800998ecf8427e
\ No newline at end of file
+5ff6163c7004300edb5e0c6670e547d3
\ No newline at end of file
diff --git a/test/as02/iecdv25.md5s b/test/as02/iecdv25.md5s
index ec6da267..8e831eb3 100644
--- a/test/as02/iecdv25.md5s
+++ b/test/as02/iecdv25.md5s
@@ -1 +1 @@
-abce8e563ec3408a0dbf10ae34dfe30c;ca114c9f2707afde489d6e6066f8833c;e4a2bdff4418802e3d127654ee906dc6;99538bb806a99596b95236ac8a965b52
\ No newline at end of file
+227a7e8a6b34571cfb3a9b70642c0183;4e4605266cb9d916830ea1eef722321d;d8b021e1b8bc3bd3796676b0b5e8f594;16ff89dda611b560eda0739fd90d0214
\ No newline at end of file
diff --git a/test/avid_mxf/mpeg2lg_422p_hl_1080i.md5s b/test/avid_mxf/mpeg2lg_422p_hl_1080i.md5s
index 606895d7..8e1e6d3e 100644
--- a/test/avid_mxf/mpeg2lg_422p_hl_1080i.md5s
+++ b/test/avid_mxf/mpeg2lg_422p_hl_1080i.md5s
@@ -1 +1 @@
-f95eae8d685f5c222c7e3e9fb84b9f9a;419a18c68d4b87dfb10408653a6bc58e;805d6fd84c1ed39585053c49016f52f7
\ No newline at end of file
+f189845c35fbe0a69d36b181d85937f9;0ece8877ab256657df705de6a01a8a48;f323a74f1505abc8bf201eeef07e0089
\ No newline at end of file
diff --git a/test/avid_mxf/mpeg2lg_mp_hl_1920_1080i.md5s b/test/avid_mxf/mpeg2lg_mp_hl_1920_1080i.md5s
index 2b2e3f30..793a0a50 100644
--- a/test/avid_mxf/mpeg2lg_mp_hl_1920_1080i.md5s
+++ b/test/avid_mxf/mpeg2lg_mp_hl_1920_1080i.md5s
@@ -1 +1 @@
-93d9e5755cdcdd13000b71bcad7d55d1;a31f0491c9255156d07ec78ac265311c;f3259501cd9d48ce94f5f1aa0006d697
\ No newline at end of file
+fbd8e5365fbd58c4a4908cd2e78319f7;0ece8877ab256657df705de6a01a8a48;f323a74f1505abc8bf201eeef07e0089
\ No newline at end of file
diff --git a/test/d10_mxf/d10_40_2997.md5 b/test/d10_mxf/d10_40_2997.md5
index 20d6158e..8563ad8d 100644
--- a/test/d10_mxf/d10_40_2997.md5
+++ b/test/d10_mxf/d10_40_2997.md5
@@ -1 +1 @@
-8a256f169a5cbaa6145f95a9d008df0a
\ No newline at end of file
+6e62f6adec11065622203c7e75cc2b38
\ No newline at end of file
```

nishabfhg commented:
The md5s might have changed when I rebuilt the project using the visual studio solution. I am not sure why.

philipnbbc reviewed (approved):
All done now as the build succeeds, Thanks for your contribution!

nishabfhg commented:
That's great! Thanks for the review.
